File: osmosm7/dynamic_programming_1/2294_동전2.py
#2294_동전2.py
n,k = map(int,input().split())
coins = []
for _ in range(n):
    coins.append(int(input()))
inf = 10**10
dp = [inf] * (k+1)
dp[0] = 0
for coin in coins:
    for i in range(coin,k+1):
        dp[i] = min(dp[i - coin] + 1 , dp[i])
if dp[-1] == inf:
    print(-1)
else:
    print(dp[-1])

File: sally/dynamic_programming_1/2294_동전 2.cpp
/*
# DP
# Problem: 2294
# Memory: 2180KB
# Time: 0ms
*/
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

#define INF 1e9

int n, k;
int coin[101] = {0,};

int main(void){
    cin.tie(NULL);
    ios::sync_with_stdio(false);

    cin >> n >> k;
    vector<int> dp(k+1, INF);
    for(int i = 0; i < n; i++){
        cin >> coin[i];
        dp[coin[i]] = 1;
    }

    for(int i = 1; i <= k; i++){
        for(auto j : coin){
            if(i + j <= k && dp[i] != INF){
                dp[i + j] = min(dp[i+j], dp[i] + 1);
            }
        }
    }

    if(dp[k] == INF) cout << "-1";
    else cout << dp[k];

    return 0;
}